spring-integration-core/src/main/java/org/springframework/integration/scattergather/ScatterGatherHandler.java

```diff
@@ -190,7 +190,21 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 
 		this.messagingTemplate.send(this.scatterChannel, scatterMessage);
 
-		return gatherResultChannel.receive(this.gatherTimeout);
+		Message<?> gatherResult = gatherResultChannel.receive(this.gatherTimeout);
+		if (gatherResult != null && (requestMessageHeaders.containsKey(ORIGINAL_REPLY_CHANNEL)
```

Comment:
I think the fix is much simpler:
```
if (gatherResult != null) {
    return gatherResult.getPayload();
}
```

and `AbstractReplyProducingMessageHandler` super class will take care about proper request message headers copying. I mean that your logic is fully the same what would happen by default.
So, the problem that we return fro this `handleRequestMessage()` the whole message with some headers.
And those headers are not overridden in the super class.

Reply:
Hi @artembilan ,

AbstractReplyProducingMessageHandler will not have the context of any new headers introduced by the subprocesses.

if (gatherResult != null) {
    return gatherResult.getPayload();
}

Losing of headers produced by downstream is going to be a problem. where in if developer sets some additional infromation required by some complex flow in headers those can be lost.

Since we are forced to introduce the logic in handleRequestMessage, can we also remove the addition of the ORIGINAL_REPLY_CHANNEL, ORIGINAL_ERROR_CHANNEL & GATHER_RESULT_CHANNEL in the request message to declutter the headers in all subflows and simplify the entire logic?

If you agree to that I can handle these changes as well.

Reply:
Yeah... I see your point. I have missed the fact that we return headers produced by the sub-flow. Even if we won't need them in the end, it doesn't mean that we can't consult those headers in the `gatherer` process.
So, yes, please, go ahead with your idea: we should restore request headers, but can keep all other new headers.